[JSC] Fix parsing of private fields with Unicode start characters

rkirsling opened the pull request:
#### 5fc81f8795e87f422134be7c28ca4636b1afedf3
<pre>
[JSC] Fix parsing of private fields with Unicode start characters
<a href="https://bugs.webkit.org/show_bug.cgi?id=269831">https://bugs.webkit.org/show_bug.cgi?id=269831</a>

Reviewed by Darin Adler and Yusuke Suzuki.

Our lexer can&apos;t currently handle various class-private field names with valid Unicode start characters, like `#𐡀`.
After seeing that m_current is &apos;#&apos;, we call isSingleCharacterIdentStart(peek(1)), which complains of U16_IS_SURROGATE;
instead of trying to cut a corner, we need to get the next code point from U16_GET, just like we did for m_current.

* JSTests/test262/expectations.yaml:
Mark 24 test cases passing.

* Source/JavaScriptCore/parser/Lexer.cpp:
(JSC::Lexer&lt;T&gt;::lexWithoutClearingLineTerminator):
Use U16_GET when a private field name starts with a non-Latin1 character.

Canonical link: <a href="https://commits.webkit.org/275152@main">https://commits.webkit.org/275152@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b3ddbf520ca947f3c9be0acc02bd6df6b7f4ffa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41053 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20066 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43613 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37145 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17397 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41627 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16983 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35358 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14621 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36356 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44925 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/34506 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37238 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36668 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40419 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/40679 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38796 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17487 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/47690 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9208 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17538 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9774 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17131 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me

darinadler commented on Source/JavaScriptCore/parser/Lexer.cpp:
```diff
@@ -2519,8 +2519,18 @@ JSTokenType Lexer<T>::lexWithoutClearingLineTerminator(JSToken* tokenRecord, Opt
+        else {
+            ASSERT(m_code + 1 < m_codeEnd);
+            char32_t codePoint;
+            U16_GET(m_code + 1, 0, 0, m_codeEnd - (m_code + 1), codePoint);
```
Did you check that invalid and partial sequences cause `U16_GET` to give us a code point value that will return `false` for `isNonLatin1IdentStart`?

rkirsling replied:
My intention was to simply mimic what we do in order to get `CharacterIdentifierStart` for [the case  in which there is no preceding `#`](https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/parser/Lexer.cpp#L1926-L1938)—basically, if we _do_ find an unpaired surrogate, we expect that it therefore does not satisfy `ID_Start`.

(If we're worried about this assumption, we could mimic [`currentCodePoint()`](https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/parser/Lexer.cpp#L812) instead and produce a specific `errorCodePoint`, but presumably in that case we should also update the case above.)

darinadler commented on Source/JavaScriptCore/parser/Lexer.cpp:
```diff
@@ -2519,8 +2519,18 @@ JSTokenType Lexer<T>::lexWithoutClearingLineTerminator(JSToken* tokenRecord, Opt
-        // Otherwise, it could be a valid PrivateName.
-        if (isSingleCharacterIdentStart(next) || next == '\\') {
+
+        bool isValidPrivateName = false;
```
I suppose it’s OK to write `= false` here, but it’s not needed.

rkirsling committed:
[JSC] Fix parsing of private fields with Unicode start characters
https://bugs.webkit.org/show_bug.cgi?id=269831

Reviewed by Darin Adler and Yusuke Suzuki.

Our lexer can't currently handle various class-private field names with valid Unicode start characters, like `#𐡀`.
After seeing that m_current is '#', we call isSingleCharacterIdentStart(peek(1)), which complains of U16_IS_SURROGATE;
instead of trying to cut a corner, we need to get the next code point from U16_GET, just like we did for m_current.

* JSTests/test262/expectations.yaml:
Mark 24 test cases passing.

* Source/JavaScriptCore/parser/Lexer.cpp:
(JSC::Lexer<T>::lexWithoutClearingLineTerminator):
Use U16_GET when a private field name starts with a non-Latin1 character.

Canonical link: https://commits.webkit.org/275152@main
